### PR TITLE
studio.kx.carla

### PR DESCRIPTION
**Pull request by Bleuzen**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. [**Link**](https://github.com/falkTX/Carla/pull/1535)
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by Bleuzen** [comment]
Still WIP, because:
Carla in flatpak can't see the audio devices connected to the system. I'm not sure what is missing.

Also building with PyQT 5.15 fails, but 5.14 works for now.
If you are interested, I can also commit the 5.15 version, but have no idea how to fix it yet.

File: com.github.falkTX.Carla.json
Comment (hfiguiere) [inline com.github.falkTX.Carla.json:25]:
```suggestion
    "--env=LV2_PATH=$HOME/.lv2:/app/extensions/Plugins/lv2",
```

File: com.github.falkTX.Carla.json
Comment (hfiguiere) [inline com.github.falkTX.Carla.json:115]:
```suggestion
```
This arch is no longer support so you don't need it.

File: appdata.xml
Comment (hfiguiere) [inline appdata.xml:18]: missing `oars`, `screenshots`. so it won't validate.

Comment (hfiguiere) [reply appdata.xml:18]: also the file should be submitted upstream.

File: com.github.falkTX.Carla.json
Comment (hfiguiere) [inline com.github.falkTX.Carla.json:24]:
```suggestion
```

There is no libgig in this package.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
bot, build com.github.falkTX.Carla

File: com.github.falkTX.Carla.json
Comment (Bleuzen) [reply com.github.falkTX.Carla.json:24]: Is there a reason you only suggested this for lv2 and not also vst3?
Already saw this on Ardour: https://github.com/flathub/org.ardour.Ardour/pull/32
Is there something bad about loading vst3 plugins from home?

File: appdata.xml
Comment (Bleuzen) [reply appdata.xml]: added screenshots and submitted it upstream: https://github.com/falkTX/Carla/pull/1535
not sure about oars, since there is a xml node for it. Any additional info needed?

File: com.github.falkTX.Carla.json
Comment (Bleuzen) [reply com.github.falkTX.Carla.json]: added both lv2 and vst3 plugin paths for home now

**Comment by Bleuzen** [comment]
bot, build com.github.falkTX.Carla

**Comment by Bleuzen** [comment]
bot, build com.github.falkTX.Carla

**Comment by Bleuzen** [comment]
bot, build com.github.falkTX.Carla

File: appdata.xml
Comment (hfiguiere) [reply appdata.xml]: nevermind about oars, I missed seeing it here. It's good. Sorry abou that.

File: com.github.falkTX.Carla.json
Comment (hfiguiere) [reply com.github.falkTX.Carla.json]: Following the spec for VST I believe it's not necessary, at least for Ardour (I remember the conversation with Robin about it since VST3_PATH was ignored).

I have to check with Carla though.

Comment (hfiguiere) [reply com.github.falkTX.Carla.json]: also loading binaries from out of the sandbox is doomed to fail - even more when developers assume Linux = Ubuntu (from a few years ago) x86_64. But that's not what I was getting at here.

**Comment by Bleuzen** [comment]
Tested this build. Works as described. It can load plugins, operates as a jack client, but can't see devices and other clients.

Native Carla:
image

Carla running in flatpak:
image

Still not sure what to do about it (missing permissions?).

**Comment by hfiguiere** [comment]
do you have pipewire running on the *host* ?

**Comment by Bleuzen** [comment]
@hfiguiere Yes, using it as replacement for pulse and jack:
```
❯ pactl info
Server String: /run/user/1000/pulse/native
Library Protocol Version: 35
Server Protocol Version: 35
Is Local: yes
Client Index: 58
Tile Size: 65472
User Name: elias
Host Name: archlinux
Server Name: PulseAudio (on PipeWire 0.3.40)
Server Version: 15.0.0
Default Sample Specification: float32le 2ch 48000Hz
Default Channel Map: front-left,front-right
Default Sink: alsa_output.usb-MOTU_M2_M20000044780-00.analog-stereo
Default Source: alsa_input.usb-MOTU_M2_M20000044780-00.analog-stereo
Cookie: 8193:3e3d
```

File: com.github.falkTX.Carla.json
Comment (tinywrkb) [inline com.github.falkTX.Carla.json:18]: I don't think this can work considering the PID namespace.  
Apps should use the [Realtime Portal](https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.Realtime).  
If you added this for PipeWire (JACK) then I believe it's wrong and should be dropped.

Comment (hfiguiere) [reply com.github.falkTX.Carla.json:18]: Some app crash if they don't have access to it - speaking from experience here. 

(can't speak for this one)

File: com.github.falkTX.Carla.json
Comment (tinywrkb) [inline com.github.falkTX.Carla.json:71]: Please look at how I [packaged pyqt](https://github.com/flathub/org.qutebrowser.qutebrowser/blob/master/pyqt5/pyqt5.json), and the [configure script](https://github.com/flathub/org.qutebrowser.qutebrowser/blob/master/pyqt5/sip/sip-build-configure.in). This will help you avoid manually changing the Python version string when switching to a new runtime version.

File: com.github.falkTX.Carla.json
Comment (tinywrkb) [reply com.github.falkTX.Carla.json:18]: @hfiguiere 
Thanks for pointing this out, I haven't hit this issue yet with the few JACK apps I packaged.

Comment (hfiguiere) [reply com.github.falkTX.Carla.json:18]: maybe it got fixed since. But back when it was starting to work that was the case. And I don't really have the time to regress everything.

File: com.github.falkTX.Carla.json
Comment (hfiguiere) [reply com.github.falkTX.Carla.json:71]: maybe we should make a shared-module for this.

Comment (tinywrkb) [reply com.github.falkTX.Carla.json:71]: I started working on this, but I'm not sure what is the best way to let maintainer choose different configure options (enabled PyQt bindings) while not breaking automated updates/f-e-d-c of Flathub's shared-modules or the app packaging.

Comment (hfiguiere) [reply com.github.falkTX.Carla.json:71]: We shouldn't need to choose different configure option. You can't when using distro package.

Comment (Bleuzen) [reply com.github.falkTX.Carla.json:71]: Looking at all the files needed for pyqt5 in your flatpak, this looks way too complicated for building one simple dependency. Also I'm still a beginner to all of this, so not sure what exactly is really needed for Carla (to copy the pyqt5 build files from your project may be too much right?).

File: com.github.falkTX.Carla.json
Comment (Bleuzen) [reply com.github.falkTX.Carla.json:18]: I found this in most other jack apps, so put it here. Not sure if it is really still needed. Could try without. But it is hard to test for now because something isn't working anyway (Carla can't see other clients and devices).

Comment (Bleuzen) [reply com.github.falkTX.Carla.json]: removed it for now because it doesn't crash Carla

**Comment by hfiguiere** [comment]
bot, build com.github.falkTX.Carla

**Comment by hfiguiere** [comment]
I am not sure what's expected to be visible or not, but Carla as an audio plugin patch bay works for me. I don't think you have access to the whole graph from inside the sandbox. But with qjackctl outside the sandbox I can.

**Comment by hfiguiere** [comment]
LGTM

**Comment by hfiguiere** [comment]
short of the few changes mentionned

(yes I will update the documentation and other flatpaks)

**Comment by Bleuzen** [comment]
@hfiguiere 

> I am not sure what's expected to be visible or not

Everything should be visible. Just like in native running Carla.
My main use case for Carla is to insert plugins in the graph and route applications through them. This is not possible without Carla having access to the full graph, so in this state it is pretty much useless.

> but Carla as an audio plugin patch bay works for me

Yes but need access to individual plugins, which is not possible currently because they are all running under one Carla jack client, which is a hefty limitation.


> I don't think you have access to the whole graph from inside the sandbox.

This should be possible. I even checked it with Ardour before starting this flatpak to make sure.
Example:

Here I have Guitarix and VLC running natively on the host (NO flatpak) and also Ardour running (IN flatpak).
Now look at what Ardour as a flatpak can see:

image

Also Ardour as a flatpak can see all connected audio devices:
image


So flatpaks can see the whole graph. But I'm not sure why Carla doesn't.

File: com.github.falkTX.Carla.json
Comment (Bleuzen) [reply com.github.falkTX.Carla.json]: now that i got Carla fully working, I can confirm that this is not needed

File: com.github.falkTX.Carla.json
Comment (Bleuzen) [reply com.github.falkTX.Carla.json]: > I have to check with Carla though.

I tested it now and turns out that Carla needs this path in env to find plugins in home. However it doesn't like to "$HOME" variable. It doesn't resolve to a full path in Carla.
I have to add
`--env=VST3_PATH=/home/elias/.vst3`
for it to work and if I do:
`--env=VST3_PATH=$HOME/.vst3`
Carla won't find the plugins in my user directory.

Comment (Bleuzen) [reply com.github.falkTX.Carla.json]: > also loading binaries from out of the sandbox is doomed to fail

Not exactly. Most devs compile against LTS version of Ubuntu. This is old enough to work in Flatpak, so most plugins work just fine.

File: com.github.falkTX.Carla.json
Comment (hfiguiere) [inline com.github.falkTX.Carla.json:23]:
```suggestion
    "--env=LV2_PATH=$HOME/.lv2:/app/extensions/Plugins/lv2:/app/lib/lv2",
```

You need that one too (documentation needs update).
Also TIL that `$HOME` isn't expanded. :-(

Comment (tinywrkb) [reply com.github.falkTX.Carla.json]: > You need that one too (documentation needs update). Also TIL that `$HOME` isn't expanded. :-(

Maybe Flatpak should bind mount homedir to `/var/home` like done with the xdg-dirs to /var/{config,data,cache}.

File: com.github.falkTX.Carla.json
Comment (hfiguiere) [reply com.github.falkTX.Carla.json]: And then they link against `libcurl-gnutls.so` because why not, everybody run Ubuntu, right?

See https://github.com/flathub/org.freedesktop.LinuxAudio.BaseExtension/issues/7

File: com.github.falkTX.Carla.json
Comment (hfiguiere) [reply com.github.falkTX.Carla.json]: I filed https://github.com/flatpak/flatpak/issues/4596

File: com.github.falkTX.Carla.json
Comment (Bleuzen) [reply com.github.falkTX.Carla.json]: I will leave it at this since it is the best we can do for now.

**Comment by Bleuzen** [comment]
Removed the "[WIP]" from title because this is considered done and tested now.

There are only improvements left, namely:
- build with newer version of pyqt5, possibly use a shared-module when available (could need help from @tinywrkb)
- use app metainfo from upstream when merged

but both don't hinder Carla from working and can be done afterwards. So I think it is ready to merge now if no other regressions come up.

**Comment by barthalion** [comment]
Has upstream developer been contacted about this submission, i.e. maintenance specifically?

**Comment by Bleuzen** [comment]
@barthalion In the main post there is a pull request linked, so the developer knows about this submission. However I haven't asked about maintenance.

@falkTX Do you want to (co-)/maintain this flatpak?

**Comment by falkTX** [comment]
> @falkTX Do you want to (co-)/maintain this flatpak?

Not really, I am yet to be convinced that flatpak is a good way to do plugins. The possibility of mixing different library versions is a major issue.

Plus I don't see myself ever using flatpaks myself, so I rather not have something labelled as supported that I don't use.

**Comment by falkTX** [comment]
btw please don't use `com.github.falkTX.Carla` as id, github hosting is likely temporary.
Use `studio.kx.carla` instead, thanks.

**Comment by hfiguiere** [comment]
> Not really, I am yet to be convinced that flatpak is a good way to do plugins. The possibility of mixing different library versions is a major issue.

Which is something that Flatpak actually take care of - unless random binaries on random distro. (not trying to convince you her)

BTW this is just the standalone version of Carla. Not Carla as a plugin.

**Comment by Bleuzen** [comment]
updated now the app id

**Comment by Bleuzen** [comment]
bot, build studio.kx.carla

**Comment by barthalion** [comment]
If you don't mind @falkTX, we would like you to grant write access to the Flathub repo, nonetheless, just so you are not blocked from closing issues or merging PRs.

/merge @falkTX

**Comment by falkTX** [comment]
Ok, so we can continue conversation there. I will create tickets as action items for individual things

**Pull request closed without merging**
